internal/webhook/dscinitialization/integration_test.go

```diff
@@ -0,0 +1,86 @@
+package dscinitialization_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/rs/xid"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	dsciv1 "github.com/opendatahub-io/opendatahub-operator/v2/api/dscinitialization/v1"
+	"github.com/opendatahub-io/opendatahub-operator/v2/internal/webhook/dscinitialization"
+	"github.com/opendatahub-io/opendatahub-operator/v2/internal/webhook/shared"
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/utils/test/envt"
+
+	. "github.com/onsi/gomega"
+)
+
+// TestDSCIWebhook_Integration exercises the validating webhook logic for DSCInitialization resources.
+// It uses table-driven tests to verify singleton enforcement and deletion restrictions in a real envtest environment.
+func TestDSCIWebhook_Integration(t *testing.T) {
+	t.Parallel()
+
+	testCases := []struct {
+		name  string
+		setup func(ns string) []client.Object
+		test  func(g Gomega, ctx context.Context, k8sClient client.Client, ns string)
+	}{
+		{
+			name: "Singleton enforcement: allows creation if none exist, denies if one exists",
+			setup: func(ns string) []client.Object {
+				return nil
+			},
+			test: func(g Gomega, ctx context.Context, k8sClient client.Client, ns string) {
+				dsci1 := shared.NewDSCI("dsci-one", ns)
+				g.Expect(k8sClient.Create(ctx, dsci1)).To(Succeed(), "should allow creation of first DSCI")
+				dsci2 := shared.NewDSCI("dsci-two", ns)
+				err := k8sClient.Create(ctx, dsci2)
+				g.Expect(err).NotTo(Succeed(), "should not allow creation of a second DSCI")
+			},
+		},
+		{
+			name: "Deletion restriction: allows deletion if no DSC exists",
+			setup: func(ns string) []client.Object {
+				return []client.Object{
+					shared.NewDSCI("dsci-delete", ns),
+				}
+			},
+			test: func(g Gomega, ctx context.Context, k8sClient client.Client, ns string) {
+				dsci := &dsciv1.DSCInitialization{}
+				key := types.NamespacedName{Name: "dsci-delete", Namespace: ns}
+				g.Expect(k8sClient.Get(ctx, key, dsci)).To(Succeed(), "should get DSCI for deletion test")
+				g.Expect(k8sClient.Delete(ctx, dsci)).To(Succeed(), "should allow deletion if no DSC exists")
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			t.Logf("Starting test case: %s", tc.name)
+			ctx, env, teardown := shared.SetupEnvAndClient(
+				t,
+				[]envt.RegisterWebhooksFn{
+					dscinitialization.RegisterWebhooks,
+				},
+				30*time.Second,
+			)
+			defer teardown() // Always clean up, even on panic/failure
```

Comment:
it does not change much, but you can use the `t.Cleanup` helper to register the teardown function to be execute when the tests ends